clang/include/clang/3C/ConstraintVariables.h

```diff
@@ -360,22 +362,27 @@ class PointerVariableConstraint : public ConstraintVariable {
   // D: If this constraint is generated because of a variable declaration, this
   //    should be a pointer to the the declaration AST node. May be null if the
   //    constraint is not generated by a declaration.
-  // N: Name for the constraint variable. This may be chosen arbitrarily, as it
-  //    it is only used for labeling graph nodes. Equality is based on generated
-  //    unique ids.
+  // N: Name for the constraint variable. Use for rewriting declarations that
+  //    need an identifier. Otherwise used to give descriptive names to atoms in
+  //    the constraint graph
   // I and C: Context objects required for this construction. It is expected
   //          that all constructor calls will take the same global objects here.
   // inFunc: If this variable is part of a function prototype, this string is
   //         the name of the function. nullptr otherwise.
   // ForceGenericIndex: CheckedC supports generic types (_Itype_for_any) which
   //                    need less restrictive constraints. Set >= 0 to indicate
   //                    that this variable should be considered generic.
+  // TSI: TypeSourceInfo object gives access to information about the source
+  //      code representation of the type. Allows for more precise rewriting by
+  //      preserving the exact syntax used to write types that aren't rewritten
+  //      by 3C.
   PointerVariableConstraint(const clang::QualType &QT, clang::DeclaratorDecl *D,
                             std::string N, ProgramInfo &I,
                             const clang::ASTContext &C,
                             std::string *InFunc = nullptr,
                             int ForceGenericIndex = -1,
-                            bool VarAtomForChecked = false);
+                            bool VarAtomForChecked = false,
+                            TypeSourceInfo *TSI = nullptr);
```

Comment:
This seems valuable enough that you may want to put it ahead of the other params with defaults. 

On the other hand, can it be obtained from `D`? No use adding another param in a long list if you can extract and cache it in the body.

Reply:
The reason for adding this parameter is that I need the `TypeSourceInfo` instance for typedefs, but the typdef declaration class is not a subclass of `DeclaratorDecl`. I suppose the other way to write this would be to replace `DeclaratorDecl *D` with `NamedDecl *D` (the common super class of `TypedefDecl` and `DeclaratorDecl`. I'm not sure how that would work out. Probably would need to add a few `dyn_cast` checks. It would also let this constructor take some other declaration sub-classes that we don't necessarily want it to accept.

Reply:
I see, it's a special case for typedefs when `D->getTypeSourceInfo()` doesn't provide what we want. Probably not worth reworking the function for that.

Reply:
I added a constructor that populates `TypeSourceInfo` from a typedef declaration. I think that makes the reasoning for the parameter a bit clearer.